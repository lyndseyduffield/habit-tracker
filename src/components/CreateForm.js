import React, { useState } from "react";
import useForm from "react-hook-form";
import DatePicker from "react-datepicker";
import { addHabit } from "../actions";
import { connect } from "react-redux";
import "react-datepicker/dist/react-datepicker.css";

const CreateForm = props => {
  const [state, setState] = useState({
    startDate: new Date(),
    endDate: new Date()
  });

  const { register, handleSubmit, setValue, errors } = useForm();

  React.useEffect(() => {
    register({ name: "startDate" });
    register({ name: "endDate" });
    setValue("startDate", new Date());
    setValue("endDate", new Date());
  }, [register, setValue]);

  const onSubmit = data => {
    const habit = {
      ...data,
      streak: [false]
    };
    props.dispatch(addHabit(habit));
    props.history.push("/");
  };

  const onDatePickerChange = (key, date) => {
    if (key === "startDate" && date > state.endDate) {
      setValue("startDate", date);
      setValue("endDate", date);
      setState({ startDate: date, endDate: date });
    } else {
      setValue(key, date);
      setState({ ...state, [key]: date });
    }
  };

  console.log(errors);

  // { startDate, endDate } -> Date -> Effect
  // Check if the start date is before the end date and if not then update ... ... ..

  // Rendering
  return (
    <form class="container" onSubmit={handleSubmit(onSubmit)}>
      <div class="field">
        <label class="label">Habit</label>
        <div class="control">
          <input
            class="input"
            type="text"
            name="title"
            placeholder="Your new habit"
            ref={register({ required: true })}
          />
        </div>
        <p class="help is-danger">{errors.title && "A title is required"}</p>
      </div>
      <div class="field">
        <label class="label">Goal</label>
        <div class="control">
          <textarea
            class="textarea"
            type="text"
            name="goal"
            placeholder="What is your goal?"
            ref={register}
          />
        </div>
      </div>
      <div class="field is-grouped">
        <label class="label">Start Date</label>
        <div class="control">
          <DatePicker
            class="dropdown"
            minDate={new Date()}
            selected={state.startDate}
            onChange={date => {
              onDatePickerChange("startDate", date);
            }}
          />
        </div>
        <label class="label">End Date</label>
        <div class="control">
          <DatePicker
            minDate={state.startDate}
            selected={state.endDate}
            onChange={date => {
              onDatePickerChange("endDate", date);
            }}
          />
        </div>
      </div>
      <h3 class="label is-large">Accountability Partner</h3>
      <div class="field">
        <label class="label">Name</label>
        <div class="control">
          <input
            class="input"
            type="text"
            name="accountabilityPartner.name"
            placeholder="Your accountability partner's name"
            ref={register({
              maxLength: { value: 40, message: "This name is too long" }
            })}
          />
        </div>
        <p class="help is-danger">
          {errors["accountabilityPartner.name"] ? (
            <span>{errors["accountabilityPartner.name"].message}</span>
          ) : (
            ""
          )}{" "}
        </p>
      </div>
      <div class="field">
        <label class="label">Email</label>
        <div class="control">
          <input
            class="input"
            type="text"
            name="accountabilityPartner.email"
            placeholder="Their email"
            ref={register({
              pattern: {
                value: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
                message: "Please enter a valid email"
              }
            })}
          />
        </div>
        <p class="help is-danger">
          {errors["accountabilityPartner.email"] ? (
            <span>{errors["accountabilityPartner.email"].message}</span>
          ) : (
            ""
          )}{" "}
        </p>
      </div>
      <div class="control">
        <button class="button is-link">Submit</button>
      </div>
    </form>
  );
};

export default connect()(CreateForm);
