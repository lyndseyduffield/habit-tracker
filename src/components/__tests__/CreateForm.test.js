import React from "react";
import CreateForm from "../CreateForm";
import renderer from "react-test-renderer";
import { testReducer } from "../../reducers/__tests__/index.test";
import { createStore } from "redux";
import { Provider } from "react-redux";

const store = createStore(testReducer);

it("renders the create form correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <CreateForm />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
