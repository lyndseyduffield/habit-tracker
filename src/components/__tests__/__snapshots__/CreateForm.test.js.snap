// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the create form correctly 1`] = `
<form
  class="form-container"
  onSubmit={[Function]}
>
  <div
    class="field"
  >
    <label
      class="label is-large"
    >
      Habit
    </label>
    <div
      class="control"
    >
      <input
        class="input"
        name="title"
        placeholder="Your new habit"
        type="text"
      />
    </div>
  </div>
  <div
    class="field"
  >
    <label
      class="label"
    >
      Goal
    </label>
    <div
      class="control"
    >
      <textarea
        class="textarea"
        name="goal"
        placeholder="What is your goal?"
        type="text"
      />
    </div>
  </div>
  <div
    class="field is-grouped is-horizontal"
  >
    <div
      class="field-body"
    >
      <div
        class="field"
      >
        <label
          class="label"
        >
          Start Date
        </label>
        <div
          class="control"
        >
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                className=""
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                type="text"
                value="02/04/2020"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label"
        >
          End Date
        </label>
        <div
          class="control"
        >
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                className=""
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                type="text"
                value="02/04/2020"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <h3
    class="label is-large margin-top"
  >
    Accountability Partner
  </h3>
  <div
    class="field"
  >
    <label
      class="label"
    >
      Name
    </label>
    <div
      class="control"
    >
      <input
        class="input"
        name="accountabilityPartner.name"
        placeholder="Your accountability partner's name"
        type="text"
      />
    </div>
  </div>
  <div
    class="field"
  >
    <label
      class="label"
    >
      Email
    </label>
    <div
      class="control"
    >
      <input
        class="input"
        name="accountabilityPartner.email"
        placeholder="Their email"
        type="text"
      />
    </div>
  </div>
  <div
    class="control margin-top"
  >
    <button
      class="button is-link"
    >
      Submit
    </button>
  </div>
</form>
`;
